### PR TITLE
CP-20682: Assert free vGPU capacity on receiving host for cross-pool SXM

### PR DESCRIPTION
**Pull request by robhoes**
We do this in the VGPU import handler when the `live` flag is set. The VM
import handler is called on the receiving host of a cross-pool migration, by
both VM.assert_can_migrate (as a dry-run) and VM.migrate_send (for real).

Signed-off-by: Rob Hoes <rob.hoes@citrix.com>

### CODE REVIEW

File: ocaml/xapi/import.ml [hunk 1]
```diff
@@ -234,6 +234,27 @@ let assert_can_live_import __context rpc session_id vm_record =
   if vm_record.API.vM_power_state = `Running || vm_record.API.vM_power_state = `Paused
   then assert_memory_available ()
 
+(* Assert that the local host, which is the host we are live-migrating the VM to,
+ * has free capacity on a PGPU from the given VGPU's GPU group. *)
+let assert_can_live_import_vgpu ~__context vgpu_record =
+  let host = Helpers.get_localhost ~__context in
+  let local_pgpus = Db.PGPU.get_refs_where ~__context ~expr:Db_filter_types.(And
```
Comment (mseri) [inline ocaml/xapi/import.ml]: I was unaware of this `get_..._where ~expr:...` syntax :+1:

Comment (robhoes) [inline ocaml/xapi/import.ml]: It's ugly, and not very well typed, but it works...

File: ocaml/xapi/import.ml [hunk 2]
```diff
@@ -234,6 +234,27 @@ let assert_can_live_import __context rpc session_id vm_record =
...
+    (Eq (Field "GPU_group", Literal (Ref.string_of vgpu_record.API.vGPU_GPU_group)),
```
Comment (mseri) [inline ocaml/xapi/import.ml]: The `GPU_group` of the source pool and destination pool has the same reference?

Comment (robhoes) [inline ocaml/xapi/import.ml]: No, but `vgpu_record` represents the VGPU record that we will create on the receiving pool, if all goes well. The reference fields in there have already been resolved for the receiving pool by the import handlers of the relevant objects.

Comment (mseri) [inline ocaml/xapi/import.ml]: Can you add this comment to the commit? I think it's useful for future reference

File: ocaml/xapi/import.ml (hunk 1 above)
Comment (mseri) [inline ocaml/xapi/import.ml]: I find it much nicer that using the same kind of call but serialising the expression by hand (I did that in the past)

**Pull request merged**